[Docs] Dynamic filters (#891)
mckinsey/vizro: vizro-core/docs/pages/user-guides/data.md

### Configure cache

> By default, each time the dashboard is refreshed a dynamic data function executes again. Data loading is batched so that a dynamic data function that supplies multiple graphs on the same page will only execute _once_ on page refresh. Even with this batching, if loading your data once is still a slow operation, your dashboard performance may suffer.

Reviewer: ```suggestion
By default, a dynamic data function executes every time the dashboard is refreshed. Data loading is batched so that a dynamic data function that supplies multiple graphs on the same page only executes _once_ on page refresh. Even with this batching, if loading your data is still a slow operation, your dashboard performance may suffer.
```

### Parametrize data loading

> You can supply arguments to your dynamic data loading function that can be modified from the dashboard. For example:

Reviewer 2: There seems to be an issue with the itemization in the rendered docs

Author: Good spot, fixed!

- To load different versions of the same data.

> - To handle big data you can use an argument that controls the amount of data that is loaded. This effectively pre-filters data before it reaches the Vizro dashboard.

Reviewer 3: The example below covers the second bullet point. Should we add an example for the first bullet point as well? I thought the code examples would come in the same order as the bullet points 😄

Author: Very good idea. Actually I'd never really thought of the example shown covering the second bullet point. It's more that it's the simplest possible example that I could think of, but you're right that a random sampling of points _sort of_ represents pre-filtering the data. The reason I say _sort of_ is because in reality you probably wouldn't want to randomly sample points. Instead the pre-filtering would take in `species` or a number of rows and return something deterministic based on that.

So in the end I think we could do three examples:
- simplest possible just to demonstrate basic principles, like now (if it still seems valuable given the below two)
- loading different versions of data
- handle big data

I'll leave adding any more examples for now and let's revisit later once we've got this feature fully completed since I'd like to spend some time coming up with really good examples here that can demonstrate all the different parts of the feature.

Reviewer: Sounds good. Maybe in that case you need to include a sentence to say something like "The following is a simple example to illustrate X" so it's not interpreted as one or the other bullet and also introduces the example clearly.

Author: Nice idea, done!

You cannot pass [nested parameters](parameters.md#nested-parameters) to dynamic data. You can only target the top-level arguments of the data loading function, not the nested keys in a dictionary.

### Filters

By default, when a [filter](filters.md) depends on dynamic data, the available selector values update on page refresh to reflect the latest dynamic data. The mechanism for this, including caching, works exactly like other non-control components such as `vm.Graph`. However, unlike such components, a filter can depend on multiple data sources.

> Vizro looks at the data source for each component in the filter's `targets`. If at least one of these is a dynamic data source then the filter behaves dynamically and the filter's selector updates on page refresh. Remember that when `targets` is not explicitly specified, a filter applies to all the components on a page that use a DataFrame including `column`.

Reviewer 3: Just to clarify: Suppose I have a page with charts connected to two different data sources, both of which have a year column. I add a Slider with `year` as the column. Let's say data source A is dynamic, while data source B is static.

On the initial load, the Slider will show all years from both data A and data B. Now, if data source A updates and has fewer years, then on page refresh, the Slider will display the **updated years from data A** and **all the years from data B.** Correct? It is **not the case** that now only the updated years from data A are shown, but still the union of the two.

Something like this:

```
import vizro.models as vm
import vizro.plotly.express as px
from vizro import Vizro

gapminder = px.data.gapminder().query("country == 'Nigeria' and year > 1990")
gapminder_two = px.data.gapminder().query("country == 'Nigeria' and year <= 1990")

bar = px.bar(gapminder,  x="year", y="lifeExp")
bar_two = px.bar(gapminder_two,  x="year", y="lifeExp", color_discrete_sequence=["violet"])

page = vm.Page(title="My page",
               components=[vm.Graph(figure=bar), vm.Graph(figure=bar_two)],
               controls=[vm.Filter(column="year")])

dashboard = vm.Dashboard(pages=[page])

if __name__ == "__main__":
    Vizro().build(dashboard).run()
```

Reviewer 2: In other words, it always shows the **intersection** of the the values of (A,B), but it shows the previously selected options right?

Author: Done!

> Vizro looks at the data source for each component in the filter's `targets`. If at least one of these is a dynamic data source then the filter behaves dynamically and the filter's selector updates on page refresh. Remember that when `targets` is not explicitly specified, a filter applies to all the components on a page that use a DataFrame including `column`.

Reviewer: I'm a little uncomfortable with the phrase "Vizro looks at..." could we revise to make it less anthropomorphic? Something like:

"If at least one of the data sources in the filter's `targets` is a dynamic data source, the filter behaves dynamically and the filter's selector updates on page refresh... "

Author: This is much better, thank you!

> When the page is refreshed, the behaviour of a dynamic filter is as follows:

Reviewer 2: Nitpick, but I stumbled: what is a dynamic filter, it was never defined. Of course it is a filter behaving dynamically, but maybe we could be clearer in the language

Reviewer 2: It might help to say clearly somwhere at the start: dynamic filter = no selector defined, or no options on selector. Static filter = options on selector defined?

This focusses less on the underlying data, but more on the filter. Maybe both can be described.

Reviewer 4: I found this section a bit challenging to follow. Here's the explanation:

_If you have a filter that depends on dynamic data but do not want the available values to change when the dynamic data changes then you should manually specify the `selector`'s `options` field (categorical selector) or `min` and `max` fields (numerical selector)._

I appreciate how this is presented as a feature - it highlights the ability to maintain static options and filters even when working with dynamic data sources. However, I think it would be helpful to briefly mention this benefit earlier in the paragraph or include a link to the detailed explanation below for quicker reference.

Author: All very good points, thank you! Implemented some rewording and restructuring based on this.

- The filter's selector updates its available values:
    - For [categorical selectors](selectors.md#categorical-selectors), `options` updates.
    - For [numerical selectors](selectors.md#numerical-selectors), `min` and `max` update.
- The value selected on screen by a dashboard user _does not_ change. If the selected value is not present in the new set of available values then it is still selected, but the filtering operation might result in an empty DataFrame.
- The `targets` of the filter and selector type do not change (for example, a `Dropdown` cannot turn into `RadioItems`).

For example, let us add two filters to the [dynamic data example](#dynamic-data) above:

!!! example "Dynamic filters"

    ```py hl_lines="20 21"
    from vizro import Vizro
    import pandas as pd
    import vizro.plotly.express as px
    import vizro.models as vm

    from vizro.managers import data_manager

    def load_iris_data():
        iris = pd.read_csv("iris.csv")

>         return iris.sample(50)

Reviewer 4: Does it make more sense to return 5 or 10 points so that filter changes are visible more often?

Author: Yes, done!

If you have a filter that depends on dynamic data but do not want the available values to change when the dynamic data changes then you should manually specify the `selector`'s `options` field (categorical selector) or `min` and `max` fields (numerical selector). In the above example, this could be achieved as follows:

```python title="Override selector options to make a dynamic filter static"
controls = [
    vm.Filter(column="species", selector=vm.Dropdown(options=["setosa", "versicolor", "virginica"])),
    vm.Filter(column="sepal_length", selector=vm.RangeSlider(min=4.3, max=7.9)),
]
```

If you [specify a custom selector](filters.md#changing-selectors) for a dynamic filter without manually specifying `options` (categorical selector) or `min` and `max` (numerical selector) then the selector remains dynamic. For example:

```python title="Dynamic filter with custom selector is still dynamic"
controls = [
    vm.Filter(column="species", selector=vm.Checklist()),
    vm.Filter(column="sepal_length", selector=vm.Slider()),

> ]

Author: This is much better, thank you!

```

> When Vizro initially builds a filter that depends on parametrized dynamic data loading, data is loaded using the default argument values. This data is used to perform initial validation, check which data sources contain the specified `column` (unless `targets` is explicitly specified) and determine the type of selector to use (unless `selector` is explicitly specified).

Reviewer 2: Would it make sense to also shortly mention what happens when column types or even availability change. I presume the dashboard crashes. Just an idea 🤔

Author: Yes, I was discussing what should happen here with @petar-qb earlier today. Indeed what will happen is the filter and graphs won't render, and we've decided to leave this behaviour like that for now. I will add a note to describe this.
